### PR TITLE
exec() and execstart() in ContainerManager

### PR DESCRIPTION
**Pull request by Boran**
try it out

```php
try {
  $lookfor='vanilla2';
  $container = $manager->find($lookfor);
  $execid = $manager->exec($container, ['/bin/date']);
  # other example: $execid = $manager->exec($container, ["/bin/bash", "-c", "ls /var/www/html"]);
  print_r("Exec ID= <" . $execid. ">\n");
  $result=$manager->execstart($execid);
  print_r("Result= <" . $result->__toString() . ">\n");
} catch (Exception $e) {
  echo $e->getMessage();
}
```

FYI this equivalent to doing the following on the command line directly with the docker api:

```
curl -X POST -H "Content-Type: application/json" http://127.0.0.1:2375/containers/vanilla2/exec -d '{ "AttachStdin":false,"AttachStdout":true,"AttachStderr":true, "Tty":false, "Cmd":["/bin/date"] }'

{"Id":"213f4f0abd592b302a6af43637f4422f6387b63a9759643b3e6024655099b298"}

curl -X POST -H "Content-Type: application/json" http://127.0.0.1:2375/exec/213f4f0abd592b302a6af43637f4422f6387b63a9759643b3e6024655099b298/start -d '{ "Detach":false,"Tty":false}'
```

### CODE REVIEW

**Comment by ubermuda** [comment]
Thanks @Boran, I will check this soon

**Comment by Boran** [comment]
Any chance of this getting committed? I would like to start referencing it in my application where I use the master branch..

**Comment by hacfi** [comment]
LGTM..will give this a try after reading the docs of the command.

File: src/Docker/Manager/ContainerManager.php [hunk 1]
```diff
@@ -210,6 +210,72 @@ public function run(Container $container, callable $attachCallback = null, array
...
+     * @param \Docker\Container     $container
+     * @param array    $cmd          command to run
+     * @param boolean  $attachstdin  
+     * @param boolean  $attachstdout 
+     * @param boolean  $attachstderr 
+     * @param integer  $tty  
+     *
+     * @throws \Docker\Exception\UnexpectedStatusCodeException
+     *
+     * @return ID of the executive
```
Comment (hacfi) [inline src/Docker/Manager/ContainerManager.php]: `@return string ID of the executive`

File: src/Docker/Manager/ContainerManager.php [hunk 2]
```diff
@@ -210,6 +210,72 @@ public function run(Container $container, callable $attachCallback = null, array
...
+        $response = $this->client->post(['/containers/{id}/exec', ['id' => $container->getId()]], [
+            'body'         => Json::encode($body),
+            'headers'      => ['content-type' => 'application/json'],
+        ]);
+
+        if ($response->getStatusCode() !== "201") {
+            throw UnexpectedStatusCodeException::fromResponse($response);
+        }
+
+      return $response->json()['Id'];
```
Comment (hacfi) [inline src/Docker/Manager/ContainerManager.php]: indentation misses 2 spaces

File: src/Docker/Manager/ContainerManager.php [hunk 3]
```diff
@@ -210,6 +210,72 @@ public function run(Container $container, callable $attachCallback = null, array
...
+    }
+
+    /**
+     * Start an executive defined from exec()
+     * This can be resude several times, so if the command /bin/date in defined in exec()
+     * execstart() on that ID will return a different value each time.
+     * todo: how are instances created by exec() and used by execstart() removed/cleanedup?
+     *
+     * @param string   $id       identifier from exec()
```
Comment (hacfi) [inline src/Docker/Manager/ContainerManager.php]: Should be `$execid`

Maybe rename that to $execId to be in line with the code (for example `$containerId`) !?

File: src/Docker/Manager/ContainerManager.php [hunk 4]
```diff
@@ -210,6 +210,72 @@ public function run(Container $container, callable $attachCallback = null, array
...
+     * @param boolean  $detach
+     * @param boolean  $tty  
+     *
+     * @throws \Docker\Exception\UnexpectedStatusCodeException
+     *
+     * @return Guzzle Stream
```
Comment (hacfi) [inline src/Docker/Manager/ContainerManager.php]: `@return \GuzzleHttp\Stream\Stream` so the IDE can detect it

Comment (Boran) [inline src/Docker/Manager/ContainerManager.php]: New commit with the issues above corrected. thanks.

**Comment by hacfi** [comment]
:+1: Thanks for the update!

**Comment by ubermuda** [comment]
Code looks good, thanks!

I'll play with it a bit and see how it feels before merging it

**Comment by Boran** [comment]
Yes please. A review and commit would be good while its fresh in the mind.

I tried to use this to upload a file with without sucess, see https://forums.docker.com/t/docker-exec-api-using-stdin-to-upload-a-file/748. Maybe you have an idea how I should be using stdin.

**Comment by hacfi** [comment]
@Boran I can’t access the forum currently..how do you try to upload the file exactly? Afaik you need a url and can’t upload it directly. What you could do though is `base64_encode()` the file and then run a command via exec which echos the base64 string, decodes it and write it to a file.

**Comment by Boran** [comment]
@hacfi, here is the essentials from the forum post.

A file can be uploaded to a container using the docker command line as follows
echo bar > /tmp/foo.txt
cat /tmp/foo.txt | docker exec -i vanilla2 sh -c 'cat >/tmp/dest.foo.txt'
docker exec -i vanilla2 sh -c 'cat /tmp/dest.foo.txt'

Now, how to do this via the API? 
There are two steps, create the "exec" and start it
One can create an exec instance:
$body = [
'AttachStdin' => true,
'AttachStdout' => true,
'AttachStderr' => true,
'Tty' => false,
'Cmd' => ["/bin/bash", "-c", "cat >/tmp/dest.foo.txt"]
];
$response = $client->post(['/containers/{id}/exec', ['id' => 'vanilla2']], [
'body' => Json::encode($body),
'headers' => ['content-type' => 'application/json'],
]);
$execid=$response->json()['Id'];

Next, then to start that exec, its is unclear to me how to pipe in the stdin (i.e. the contents of /tmp/foo.txt above)
    $body = ['Detach' => false, 'Tty' => $tty ];
    $response = $client->post(['/exec/{id}/start', ['id' => $execid]], [
        'body'         => Json::encode($body),
        'headers'      => ['content-type' => 'application/json'],
    ]);
Reference doc is here: http://docs.docker.com/reference/api/docker_remote_api_v1.16/#exec-start1
It talks about a stream result (for stdout/err), but how does one stream in stdin?

**Comment by hacfi** [comment]
@Boran I’m trying to understand how the docker cli makes it work but don’t fully understand it and it’s so difficult to debug this. Have a look at https://github.com/docker/docker/blob/master/api/client/commands.go#L2503-2611 ..that’s how they do it.

Why not use base64_encode, echo, decode and write it?

I think there are still working on a feature to make this easier: https://github.com/docker/docker/issues/5846

**Comment by Boran** [comment]
Well I have not even got around to writing a helloworld() in go yet. But you are right, once has to remember to refer back to the docker sources (and I do forget that :-) )

First, in the docker API for /exec/id/start, the body  contains  ['Detach' => $detach, 'Tty' => $tty ]. So to paste in a file there, one would have to be some additional entry in that array. (or detach/tty should be URL params and the body is tied to stdin).

Next, looking at the runconfig go code you link to, stuff like detach/tty etc are parameters and not part of stdin (or the body)
https://github.com/docker/docker/blob/master/api/client/commands.go#L2571 
if _, _, err := readBody(cli.call("POST", "/exec/"+execID+"/start", execConfig, false)); err != nil
...
in = cli.in
..
errCh = promise.Go(func() error {
		return cli.hijack("POST", "/exec/"+execID+"/start", execConfig.Tty, in, out, stderr, hijacked, execConfig)
	})

https://github.com/docker/docker/blob/1dc1b9345109549e619f207f4ecbc30605b4ebb6/api/client/hijack.go#L130
..
req, err := http.NewRequest(method, fmt.Sprintf("/v%s%s", api.APIVERSION, path), params)
clientconn.Do(req)

So after the initial http request/response, stdin is then redirected? Hmm. Not sure I get this.

**Comment by hacfi** [comment]
@Boran My guess is that the go cli command processes stdin before sending it to the Docker host.

**Comment by joelwurtz** [comment]
Actually docker hijack the HTTP protocol to send the stdin into a stream for the daemon (see https://github.com/docker/docker/blob/master/api/client/hijack.go) This is actually not supported in docker-php and require too much work for me in the moment

(if you want to implement this hijack we will be able to support more things)

@Boran for your use case there is no way to do this for the moment, but you have several workaround : 

1. Add the file during the build of the image
2. Share a volume at the runtime of the container and copy the file inside it when you need
3. Have a process running in the container which can import the file (sshd with scp, ftp, ....)

Also i take the liberty to take your PR in #99, i have add test and the possibilty to add a callback for streaming the output of the exec, so i will close this PR replaced by #99

**Pull request closed without merging**